source/SH3/camera/camera.cpp

```diff
@@ -0,0 +1,131 @@
+/** @file
+ *
+ *  Implementation of camera.hpp
+ *
+ *  @copyright 2017 Palm Studios
+ *
+ *  @date 1-9-2017
+ *
+ *  @author Jesse Buhagiar
+ */
+#include <cmath>
+#include "SH3/camera/camera.hpp"
+#include "glm/gtc/matrix_transform.hpp"
+#include <algorithm>
+#include <boost/algorithm/clamp.hpp>
+
+static constexpr glm::vec3 worldUp = glm::vec3(0, 1, 0); /**< Which axis is considered 'world up'. */
+
+sh3::camera::Camera::Camera(const glm::vec3& pos, angle<float> fov, float aspect, float zNear, float zFar)
+: camPos(pos), camFov(fov), aRatio(aspect), camNear(zNear), camFar(zFar),
+  camFront(glm::vec3(0.0f, 0.0f, -1.0f)), camUp(glm::vec3(0.0f, 1.0f, 0.0f)),
+  pMatrix(glm::perspective(0.0f, aspect, zNear, zFar)),
+  vMatrix(glm::lookAt(camPos, camPos + camFront, camUp)),
+  camPitch(angle<float>::FromDegrees(-90.0f)),
+  camYaw(angle<float>::FromDegrees(0.0f)),
+  camMode(MODE::FIRST_PERSON)
+{
+    RecalculatePerspective();
+}
+
+void sh3::camera::Camera::RecalculatePerspective()
+{
+    glm::vec3 front;
+    front.x = cos(camPitch.AsRadians()) * cos(camYaw.AsRadians());
+    front.y = sin(camPitch.AsRadians());
+    front.z = cos(camPitch.AsRadians()) * sin(camYaw.AsRadians());
+    camFront = glm::normalize(front);
+
+    camRight = glm::normalize(glm::cross(camFront, worldUp));
+    camUp = glm::normalize(glm::cross(camRight, camFront));
+
+    vMatrix = glm::lookAt(camPos, camPos + camFront, camUp);
+    pMatrix = glm::perspective(camFov.AsRadians(), aRatio, camNear, camFar); // Set our perspective matrix
+}
+
+glm::mat4 sh3::camera::Camera::GetProjectionMatrix() const
+{
+    return pMatrix * glm::lookAt(camPos, camPos + camFront, camUp);
+}
+
+void sh3::camera::Camera::SetPosition(const glm::vec3& pos)
+{
+    camPos = pos;
+}
+
+void sh3::camera::Camera::Translate(const glm::vec3& trans)
+{
+    if(camMode == MODE::FIRST_PERSON)
+    {
+        camPos += trans;
+        RecalculatePerspective();
+    }
+}
+
+void sh3::camera::Camera::Translate(const float factor)
+{
+    if(camMode == MODE::FIRST_PERSON)
+    {
+        camPos += camFront * factor;
+        RecalculatePerspective();
+    }
+}
+
+void sh3::camera::Camera::SetFOV(angle<float> fov)
+{
+    camFov = fov;
+    RecalculatePerspective();
+}
+
+void sh3::camera::Camera::SetMode(sh3::camera::MODE mode)
+{
+    camMode = mode;
+}
+
+void sh3::camera::Camera::AddYaw(angle<float> yaw)
+{
+    camYaw += yaw;
+    RecalculatePerspective();
+}
+
+void sh3::camera::Camera::AddPitch(angle<float> pitch)
+{
+    camPitch = boost::algorithm::clamp(camPitch + pitch, angle<float>::FromDegrees(-89.0f), angle<float>::FromDegrees(89.0));
```

Comment:
I was having REALLY weird perspective 'flipping' issues when clamping to [-90, 90], so 89 will do for now (FPS mode isn't really important right now anyway.)

Reply:
Yeah, we can look at it later. Add a `FIXME` though. I think this might have something to do with calculating the right vector (`camRight = camFront x worldUp`); add this as a note to the `FIXME`.
That second `89.0` should be a `89.0f`.